student_slam_helper.py

```diff
@@ -1,418 +1,507 @@
-            :param kp_to_measurement: a function which computes a range/bearing measurement from the center of the
-                                      camera frame to kp
-            :return: the updated landmark
+        Return the drone's estimated position using the expectation of the
+        belief distribution, which is the weighted sum of the poses of all the
+        particles in the filter, weighted by the weight of the particle. Note
+        that the weights must be normalized to 1 for this to work.
+
+        eg: the estimated x position of the drone is the sum over all of the
+              particles of the product of the normalized weight of the
+              particle and the particle's x pose
         """
 
-        particle.weight = PROB_THRESHOLD
+        #######################################################################
+        # TODO implement this method
 
-        ##############################################################################
-        # TODO implement this method to update the map for FastSLAM! 
-        #
-        # if this particle has no landmarks:
-        #    add every currently observed feature as a landmark
-        #    adjust particle's weight to reflect the change in certainty from adding landmarks
-        # else:
-        #    close landmarks = list of particle's landmarks within self.perceptual_range of the particle's pose
-        #    if close landmarks is not empty:
-        #        part_descriptors = list of landmark.des for each landmark in close_landmarks
-        #        matched_landmarks = boolean list with same length as close_landmarks, set to false
-        #
-        #    for each observed feature:
-        #        if part_descriptors is not empty:
-        #            match = self.matcher.knnMatch(np.array([des]), np.array(part_descriptors), k=2)
-        #
-        #        if there was not match or the match is poor quality:
-        #            add the feature as a new landmark
-        #            adjust particle weight to reflect the change in certainty
-        #        else:
-        #            use match.trainIdx to find the index of the matching landmark in the map
-        #            set the boolean corresponding to this landmark in matched_landmarks to true
-        #            update the landmark in particle.landmarks with information from the newly
-        #               observed feature*
-        #            update the weight of the particle to reflect the quality of the match
-        #               hint: use scale_weight
-        #
-        # * this step is the most tricky because the index of the landmark in the close_landmarks list
-        #   is not the same as its index in particle.landmarks! This is because close_landmarks only
-        #   contains some of the landmarks in particle.landmarks
-        ##############################################################################
```

Comment:
This kind of pseudocode could be helpful for students, I would keep it otherwise it's a lot of work to think about everything from scratch.